examples/gosherve-demo/src/charm.py

```diff
@@ -0,0 +1,125 @@
+        """Validate the configuration."""
+        if not self.redirect_map:
+            raise ValueError('redirect-map must be set in the configuration.')
+        if not self.redirect_map.startswith('http'):
+            raise ValueError(
+                f'Invalid redirect-map URL: {self.redirect_map}. '
+                "It must start with 'http' or 'https'."
+            )
+
+
+class HelloKubeconCharm(ops.CharmBase):
+    """Charm the service."""
+
+    def __init__(self, framework: ops.Framework):
+        super().__init__(framework)
+        framework.observe(self.on.install, self._on_install)
+        framework.observe(self.on.config_changed, self._on_config_changed)
+        framework.observe(self.on['gosherve'].pebble_ready, self._on_config_changed)
+        framework.observe(self.on['pull_site'].action, self._pull_site_action)
+        self.ingress = IngressPerAppRequirer(
+            self,
+            port=8080,
+            host=f'{self.app.name}-endpoints.{self.model.name}.svc.cluster.local',
+            strip_prefix=True,
+        )
+
+    def _on_install(self, _: ops.InstallEvent):
+        # Download the site:
+        self._fetch_site()
+
+    def _on_config_changed(self, _: ops.ConfigChangedEvent | ops.PebbleReadyEvent):
+        """Handle the config-changed and pebble-ready events."""
+        # Get the gosherve container so we can configure/manipulate it:
+        container = self.unit.get_container('gosherve')
+        # Create a new config layer:
+        layer = self._gosherve_layer()
+
+        try:
+            container.add_layer('gosherve', layer, combine=True)
+            logging.info("Added updated layer 'gosherve' to Pebble plan.")
+            # Tell Pebble to load the new layer and restart the service if needed:
+            container.replan()
+            logging.info('Replanned gosherve service.')
+        except ops.pebble.ConnectionError:
+            self.unit.status = ops.WaitingStatus('waiting for Pebble in workload container')
+            # We assume here that we'll get a pebble-ready event later, and we
+            # will add the layer then.
+        else:
+            # All is well, set an ActiveStatus
+            self.unit.status = ops.ActiveStatus()
+
+    def _gosherve_layer(self) -> ops.pebble.LayerDict:
+        """Returns a Pebble configuration layer for Gosherve."""
+        config = self.load_config(GosherveConfig, errors='blocked')
+        return {
+            'summary': 'gosherve layer',
+            'description': 'pebble config layer for gosherve',
+            'services': {
+                'gosherve': {
+                    'override': 'replace',
+                    'summary': 'gosherve',
+                    'command': '/gosherve',
+                    'startup': 'enabled',
+                    'environment': {
+                        'REDIRECT_MAP_URL': config.redirect_map,
+                        'WEBROOT': '/srv',
+                    },
+                }
+            },
```

Comment:
We're relying on K8s resolving the service name to all units' ip addresses in IngressPerAppRequirer.

Which means that K8s effectively uses readiness checks to route or not route requests to specific units.

Thus, ideally we'd specify the readiness check... unless those have special meaning in Juju or we can't control those through  Pebble.